ios/brave-ios/Sources/Brave/Frontend/UserContent/UserScripts/__firefox__.js

```diff
@@ -412,14 +410,24 @@ if (!window.__firefox__) {
       return Promise.reject(new TypeError("undefined is not an object (evaluating 'webkit.messageHandlers')"));
     }
 
-    let webkit = window.webkit;
-    delete window.webkit.messageHandlers[messageHandlerName].postMessage;
-    delete window.webkit.messageHandlers[messageHandlerName];
-    delete window.webkit.messageHandlers;
-    delete window.webkit;
-    let result = $MessageHandlers[messageHandlerName].postMessage(message);
-    window.webkit = webkit;
-    return result;
+    return new Promise((resolve, reject) => {
+      var oldWebkit = window.webkit;
+      delete window['webkit'];
+      
+      // WebKit no longer restores the handler immediately! So we poll for when that happens and resolve the promise accordingly.
+      const timeout = 5000;
```

Comment:
A few concerns I have.
1) Does it wait 5 seconds before trying to `postMessage`? Isn't it a lot for some use cases?
2) Also a web page race during this time out and modify `window.webkit` once it is restored (or before it is restored, I'm not sure what happens in this case)

Reply:
1. Fixed.
2. Fixed.

I tested with:

```javascript
let interval = setInterval(function() {
  console.log("FOO");
}, 0);

setTimeout(function() {
   let startTime = Date.now();
    while(true) {
        console.log("NICE");
        if (Date.now() - startTime >= 5000) {
          clearInterval(interval);
          break;
        }
    }
}, 0);
```

To see if the loop will `block` the `setInterval`. So even if a page tries to replace the `window.webkit` inside of `setInterval` or something similar, the while loop is synchronous and it blocks ALL other operations (async included).

Only once the while loop completes, everything else executes :) so I changed the code to use a while loop with a timeout. I checked performance and there's no issue as webkit replaces it internally asynchronously, in the next execution frame, while the loop blocks the page from executing anything else.